### PR TITLE
dibs: I will implement exercise custom-set

### LINKED ISSUE #746: custom-set: implement exercise
There are exercises that are not yet implemented in Python track (the full list of them - https://github.com/exercism/python/issues/417#issuecomment-334841778).
You can help us change this situation by implementing this one. Ref #417

The guide ["Implement an exercise from specification"](https://github.com/exercism/docs/blob/master/you-can-help/implement-an-exercise-from-specification.md) will help you succeed.

If you still have questions, I encourage you to ask them before starting working on the issue.

I also suggest claiming the issue by creating an empty PR as stated [here](https://github.com/exercism/docs/blob/master/you-can-help/implement-an-exercise-from-specification.md#avoiding-duplicate-work)

Please note that there are some special ['closing issue' keywords](https://help.github.com/articles/closing-issues-using-keywords/) that you can put in description of your PR and Github will take care of closing the corresponding issue automatically after merge.
This simple thing helps us a lot, and makes our life a bit easier

### PR DESCRIPTION
**Pull request by WaldoHatesYou**
I will implement custom sets in [Custom Set - X-Common](https://github.com/exercism/x-common/tree/master/exercises/custom-set)

-----
Fixes #746

### CODE REVIEW

**Comment by behrtam** [comment]
If you have any questions don't hesitate to ask or commit non working code so we can have a look and help you.

File: exercises/custom-set/example.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+        for number in numbers:
+            self.add(number)                                                                                                                                                                                                                                                                                                                                                                              
+
+    def empty(self):
+        return self.size == 0
+    
+    def __len__(self):
+        return self.size
+
+    def expand(self):
```
Comment (behrtam) [inline exercises/custom-set/example.py]: I think the hole `capacity` makes this implementation too complicated and is not needed in the set context of this exercise.

File: exercises/custom-set/custom_set_test.py [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+        self.assertEqual(len(empty), 0)
+        self.assertEqual(len(non_empty), 3)
+        self.assertEqual(len(repeated), 3)
+    
+    def test_list(self):
+        empty = CustomSet()
+        non_empty = CustomSet(1,2,3)
+        repeated = CustomSet(1,3,2,1,3,2)
+
+        self.assertEqual(list(empty), [])
```
Comment (behrtam) [inline exercises/custom-set/custom_set_test.py]: Those asserts should be indented of the item order.

File: exercises/custom-set/custom_set_test.py [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+        self.assertFalse(seq.disjoint(odds))
+    
+    def test_sets_that_share_none_disjoint(self):
+        seq = CustomSet(1,2,3)
+        after = CustomSet(4,5,6)
+
+        self.assertTrue(seq.disjoint(after))
+    
+    def test_set_with_same_elems_equal(self):
+        seq = CustomSet(1,2,3)
```
Comment (behrtam) [inline exercises/custom-set/custom_set_test.py]: You are not testing two sets with the same elements just that the exact same object is equal.

**Comment by behrtam** [comment]
Would you mind fixing all the style errors. Just run `flake8` locally or check the `travis-build`: https://github.com/exercism/python#code-style

I think you are missing some of the common test cases: https://github.com/exercism/problem-specifications/blob/master/exercises/custom-set/canonical-data.json

File: exercises/custom-set/custom_set_test.py [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+        not_empty = CustomSet(1)
+        self.assertFalse(not_empty.empty())
+    
+    def test_empty_contains_nothing(self):
+        empty = CustomSet()
+        self.assertFalse(1 in empty)
+    
+    def test_simple_contains(self):
+        simple = CustomSet(1,2,3)
+        self.assertTrue(1 in simple)
```
Comment (behrtam) [inline exercises/custom-set/custom_set_test.py]: I like that you use the `in` operator and not another class method.

File: exercises/custom-set/example.py [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+                result.add(element)
+        
+        return result
+
+    def __iter__(self):
+        result = []
+        for bucket in self.data:
+            for element in bucket:
+                result.append(element)
+        return iter(result)
```
Comment (behrtam) [inline exercises/custom-set/example.py]: Why not just `yield` the elements instead of creating a copy?

File: exercises/custom-set/example.py [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+        if self.size > (2/3) * self.capacity:
+            self.expand()
+        index = hash(n) % len(self.data)
+        self.data[index].append(n)
+        self.size += 1
+        
+    def __contains__(self, n):
+        return n in self.data[hash(n) % len(self.data)]
+
+    def subset(self, other):
```
Comment (behrtam) [inline exercises/custom-set/example.py]: You could reuse your `__contains__` method like this:
```python
def subset(self, other):
    return all(map(self.__contains__, other))
```

File: exercises/custom-set/example.py [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+        for bucket in other.data:
+            for element in bucket:
+                if element not in self:
+                    return False
+        return True
+
+    def disjoint(self, other):
+        disj = CustomSet()
```
Comment (behrtam) [inline exercises/custom-set/example.py]: You could reuse your `__contains__` method like this:
```python
def subset(self, other):
    return not any(map(self.__contains__, other))
```

File: exercises/custom-set/example.py (hunk 1 above)
Comment (WaldoHatesYou) [inline exercises/custom-set/example.py]: I don't understand what you mean by this.

File: exercises/custom-set/custom_set_test.py (hunk 2 above)
Comment (WaldoHatesYou) [inline exercises/custom-set/custom_set_test.py]: What do you mean by indented of the item order?

File: config.json [hunk 8]
```diff
@@ -441,6 +441,12 @@
       ]
     },
     {
+      "slug": "custom-set",
+      "difficulty": 2,
```
Comment (m-a-ge) [inline config.json]: `uuid` is required

can be generated using `configlet`

**Comment by m-a-ge** [comment]
@WaldoHatesYou are still interested in finish this?

**Comment by cmccandless** [comment]
As this has been abandoned for some time with no response from the author, this PR will be closed.

**Pull request closed without merging**